holidays/countries/china.py

```diff
@@ -9,53 +9,361 @@
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2006年部分节假日安排的通知 # 2006
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2005年部分节假日安排的通知 # 2005
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2004年部分节假日安排的通知 # 2004
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2003年部分节假日休息安排的通知 # 2003
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2002年部分节假日休息安排的通知 # 2002
+    - https://zh.wikisource.org/wiki/国务院办公厅关于2001年春节、“五一”、“十一”放假安排的通知 # 2001
+
+    Checked With:
+    - https://www.officeholidays.com/countries/china/2023
+    - https://www.china-briefing.com/news/china-public-holiday-2023-schedule/
+    - https://www.timeanddate.com/calendar/?year=2023&country=41
+    - https://m.wannianli.tianqi.com/fangjiaanpai/2001.html # 2001-2010
+
+    Limitations:
+
+    - Only checked with the official General Office of the State Council Notice from 2001 onwards.
+
+    - Due to its complexity, need yearly checks 3-weeks before year's end each year.
     """
 
     country = "CN"
+    supported_categories = {PUBLIC, HALF_DAY}
+    default_language = "zh_CN"
+    supported_languages = ("en_US", "th", "zh_CN", "zh_TW")
+
+    # Special Cases.
+
+    # New Year's Day.
+    new_years_day_overflow = tr("元旦")
+
+    # National Day.
+    national_day_2008_golden_week = tr("国庆节")
+
+    # Mid-Autumn Festival.
+    mid_autumn_festival_2010_special = tr("中秋节")
+
+    special_public_holidays = {
+        2007: (
+            # 2007: Overflow from 2008 Notice.
+            (DEC, 30, new_years_day_overflow),
+            (DEC, 31, new_years_day_overflow),
+        ),
+        2008: (
+            # 2008: weird National Day Golden Week pattern.
+            (SEP, 29, national_day_2008_golden_week),
+            (SEP, 30, national_day_2008_golden_week),
+            (OCT, 4, national_day_2008_golden_week),
+            (OCT, 5, national_day_2008_golden_week),
+        ),
+        2010: (
+            # 2010: doesn't fit with existing observed pattern.
+            (SEP, 23, mid_autumn_festival_2010_special),
+            (SEP, 24, mid_autumn_festival_2010_special),
+        ),
+        2013: (
+            # 2013: doesn't fit with existing observed pattern.
+            (JAN, 2, new_years_day_overflow),
+            (JAN, 3, new_years_day_overflow),
+        ),
+        2018: (
+            # 2018: Overflow from 2019 Notice.
+            (DEC, 30, new_years_day_overflow),
+            (DEC, 31, new_years_day_overflow),
+        ),
+        2022: (
+            # 2022: Overflow from 2023 Notice.
+            (DEC, 31, new_years_day_overflow),
+        ),
+        2023: (
+            # 2023: Overflow from 2023 Notice.
+            (JAN, 2, new_years_day_overflow),
+        ),
+    }
 
     def __init__(self, *args, **kwargs):
         ChineseCalendarHolidays.__init__(self)
         InternationalHolidays.__init__(self)
         super().__init__(*args, **kwargs)
 
-    def _populate(self, year):
-        if year <= 1949:
+    def _add_observed_one_off_holidays(self, dt: date):
```

Comment:
It's a great function, I like this kind of 😋. But as far as I understand, China uses "substituted holidays" - moving working day to another (non-working) day to create a long continuous holidays.
For example, for 2002 (machine translation from [source](https://zh.wikisource.org/wiki/%E5%9B%BD%E5%8A%A1%E9%99%A2%E5%8A%9E%E5%85%AC%E5%8E%85%E5%85%B3%E4%BA%8E2002%E5%B9%B4%E9%83%A8%E5%88%86%E8%8A%82%E5%81%87%E6%97%A5%E4%BC%91%E6%81%AF%E5%AE%89%E6%8E%92%E7%9A%84%E9%80%9A%E7%9F%A5)):

> First, New Year's Day: January 1 -3 days off, a total of three days. Among them, 1 is a legal holiday, will be December 29, 2001 (Saturday), 30 (Sunday) of the public holiday were transferred to January 2, 2002 (Wednesday), January 3 (Thursday) off.

In our terms, it will be

```python
    substituted_holidays = {
        // ...
        2002: (
            (2001, DEC, 29, JAN, 2),
            (2001, DEC, 30, JAN, 3),
            // ...
        ),
    }
```
And so on.

Reply:
But we have to modify the `substituted_holidays` handling so that holiday name can be set, without `substituted_label` using.

Reply:
IMO, we could also probably bypass this by simply assigning `"%s"` for `substituted_label`; it's not exactly ideal, but this could work for now.

In practice, though, this would mean that we need to assign the substituted holidays manually from 2001 to 2023, which would be a pain to implement though I'm fine with that.

Reply:
> IMO, we could also probably bypass this by simply assigning `"%s"` for `substituted_label`; it's not exactly ideal, but this could work for now.

Good idea! More flexible scheme will be needed when we find country with both types of substituted holidays. 🙄

> In practice, though, this would mean that we need to assign the substituted holidays manually from 2001 to 2023, which would be a pain to implement though I'm fine with that.

I know, I did it for Ukraine and Belarus. :)) But at the same time, it's a bit interesting.

Reply:
Please let me know once you've updated the `substituted_holidays` in that case 🙏👀

Reply:
I'm merging this assuming we'll replace the observed one off method using substituted holidays approach and deal with the incomplete coverage (if any) separately.